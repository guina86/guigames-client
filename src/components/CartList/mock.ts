export default [
  {
    id: '1',
    img: 'https://res.cloudinary.com/guinacloud/image/upload/v1659445472/thumbnail_red_dead_img_a0c18e2390.jpg',
    title: 'Red Dead Redemption 2',
    price: '$215.00'
  },
  {
    id: '2',
    img: 'https://res.cloudinary.com/guinacloud/image/upload/v1659445472/thumbnail_borderlands_3_0021705309.png',
    title: 'Borderlands 3',
    price: '$215.00'
  }
]
