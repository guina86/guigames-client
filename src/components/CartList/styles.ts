import { tint } from 'polished'
import styled, { css } from 'styled-components'

export const Wrapper = styled.main`
  ${({ theme }) => css`
    background: ${theme.colors.white};
    display: flex;
    flex-direction: column;
    align-self: flex-start;
  `}
`

export const Footer = styled.div`
  ${({ theme }) => css`
    background: ${tint(0.2, theme.colors.lightGray)};
    color: ${theme.colors.black};
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.small};
    padding: 2rem;
    display: flex;
    justify-content: space-between;
    align-items: center;
  `}
`

export const Total = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.primary};
  `}
`
