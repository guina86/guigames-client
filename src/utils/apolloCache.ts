import { InMemoryCache } from '@apollo/client'
import { concatPagination } from '@apollo/client/utilities'

export const makeApolloCache = () =>
  new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          games: concatPagination(['where', 'sort'])
        }
      }
    }
  })
