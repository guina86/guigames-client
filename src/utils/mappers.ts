import { GetGames_games } from 'graphql/generated/GetGames'
import { GetHome_banners, GetHome_sections_newGames_highlight } from 'graphql/generated/GetHome'

export const imageMapper = (url: string | null | undefined) =>
  url ? `http://localhost:1337${url}` : ''

export const bannerMapper = (banners: GetHome_banners[]) =>
  banners.map((banner) => ({
    img: imageMapper(banner.image?.url!),
    title: banner.title,
    subtitle: banner.subtitle,
    buttonLabel: banner.button?.label,
    buttonLink: banner.button?.link,
    ...(banner.ribbon && {
      ribbon: banner.ribbon.text,
      ribbonColor: banner.ribbon.color,
      ribbonSize: banner.ribbon.size
    })
  }))

export const gamesMapper = (games: GetGames_games[] | null | undefined) =>
  games &&
  games.map((game?) => ({
    title: game.name,
    slug: game.slug,
    developer: game.developers[0].name,
    img: imageMapper(game.cover?.url!),
    price: game.price
  }))

export const highlightMapper = (
  highlight: GetHome_sections_newGames_highlight | null | undefined
) =>
  highlight && {
    title: highlight.title,
    subtitle: highlight.subtitle,
    backgroundImage: imageMapper(highlight.background?.url!),
    floatImage: imageMapper(highlight.floatImage?.url!),
    alignment: highlight.alignment,
    buttonLabel: highlight.buttonLabel,
    buttonLink: highlight.buttonLink
  }
