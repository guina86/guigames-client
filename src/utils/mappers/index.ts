import { GetGames_games } from 'graphql/generated/GetGames'
import { GetHome_banners, GetHome_sections_newGames_highlight } from 'graphql/generated/GetHome'
import { GetWishlist_wishlists_games } from 'graphql/generated/GetWishlist'
import { GetOrders_orders } from 'graphql/generated/GetOrders'
import formatPrice from 'utils/format-price'

export const imageMapper = (url: string | null | undefined) =>
  url ? `http://localhost:1337${url}` : ''

export const bannerMapper = (banners: GetHome_banners[]) =>
  banners.map((banner) => ({
    img: imageMapper(banner.image?.url!),
    title: banner.title,
    subtitle: banner.subtitle,
    buttonLabel: banner.button?.label,
    buttonLink: banner.button?.link,
    ...(banner.ribbon && {
      ribbon: banner.ribbon.text,
      ribbonColor: banner.ribbon.color,
      ribbonSize: banner.ribbon.size
    })
  }))

export const gamesMapper = (
  games: GetGames_games[] | GetWishlist_wishlists_games[] | null | undefined
) =>
  games
    ? games.map((game?) => ({
        id: game.id,
        title: game.name,
        slug: game.slug,
        developer: game.developers[0].name,
        img: imageMapper(game.cover?.url!),
        price: game.price
      }))
    : []

export const highlightMapper = (
  highlight: GetHome_sections_newGames_highlight | null | undefined
) =>
  highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: imageMapper(highlight.background?.url!),
        floatImage: imageMapper(highlight.floatImage?.url!),
        alignment: highlight.alignment,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink
      }
    : {}

export const cartMapper = (games: GetGames_games[] | null | undefined) => {
  return games
    ? games.map((game) => ({
        id: game.id,
        img: imageMapper(game.cover?.url),
        price: formatPrice(game.price),
        title: game.name
      }))
    : []
}

export const ordersMapper = (orders: GetOrders_orders[] | null) => {
  return orders
    ? orders.map((order) => {
        return {
          id: order.id,
          paymentInfo: {
            flag: order.card_brand,
            img: order.card_brand ? `/img/cards/${order.card_brand}.png` : null,
            number: order.card_last4 ? `**** **** **** ${order.card_last4}` : 'Free Game',
            purchaseDate: `Purchase made on ${new Intl.DateTimeFormat('en-US', {
              day: 'numeric',
              month: 'short',
              year: 'numeric'
            }).format(new Date(order.created_at))}`
          },
          games: order.games.map((game) => ({
            id: game.id,
            title: game.name,
            downloadLink: 'https://guigames.com/game/download/yuYT56Tgh43LkjhNBgdf',
            img: imageMapper(game.cover?.url),
            price: formatPrice(game.price)
          }))
        }
      })
    : []
}
