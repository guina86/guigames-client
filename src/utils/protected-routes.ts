import { GetServerSidePropsContext } from 'next'
import { getSession } from 'next-auth/client'
import { Session } from './apollo'

async function protectedRoutes(context: GetServerSidePropsContext) {
  const session = await getSession(context)

  if (!session) {
    context.res.setHeader('Location', `/sign-in?callbackUrl=${context.resolvedUrl}`)
    context.res.statusCode = 302
  }

  return session as Session
}

export default protectedRoutes
