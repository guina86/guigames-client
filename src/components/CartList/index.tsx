import Button from 'components/Button'
import Empty from 'components/Empty'
import GameItem from 'components/GameItem'
import Loading from 'components/Loading'
import { useCart } from 'hooks/use-cart'
import Link from 'next/link'
import * as S from './styles'

export type CartListProps = {
  hasButton?: boolean
  hasLinks?: boolean
}

const CartList = ({ hasButton = false, hasLinks = true }: CartListProps) => {
  const { items, total, loading } = useCart()

  if (loading)
    return (
      <S.LoadingWrapper>
        <Loading />
      </S.LoadingWrapper>
    )

  return (
    <S.Wrapper isEmpty={items.length === 0}>
      {items.length > 0 ? (
        <>
          <S.GamesList>
            {items.map((item) => (
              <GameItem key={item.title} {...item} hasLink={hasLinks} />
            ))}
          </S.GamesList>
          <S.Footer>
            {!hasButton && <span>Total</span>}
            <S.Total>{total}</S.Total>

            {hasButton && (
              <Link href="/cart" passHref>
                <Button as="a">Open cart</Button>
              </Link>
            )}
          </S.Footer>
        </>
      ) : (
        <Empty
          title="Your cart is empty"
          description="Go back to the store and explore great games and offers."
        />
      )}
    </S.Wrapper>
  )
}

export default CartList
