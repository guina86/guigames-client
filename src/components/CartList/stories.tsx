import { Story, Meta } from '@storybook/react'
import CartList, { CartListProps } from '.'
import itemsMock from './mock'

export default {
  title: 'CartList',
  component: CartList,
  args: {
    items: itemsMock,
    total: 'R$ 430,00'
  },
  argTypes: {
    items: {
      table: {
        disable: true
      }
    }
  },
  parameters: {
    backgrounds: {
      default: 'gui-dark'
    }
  }
} as Meta

export const Default: Story<CartListProps> = (args) => (
  <div style={{ maxWidth: 800 }}>
    <CartList {...args} />
  </div>
)

export const WithButton: Story<CartListProps> = (args) => (
  <div style={{ maxWidth: 800 }}>
    <CartList hasButton {...args} />
  </div>
)

export const Empty: Story<CartListProps> = (args) => (
  <div style={{ maxWidth: 800 }}>
    <CartList {...args} hasButton items={[]} />
  </div>
)
