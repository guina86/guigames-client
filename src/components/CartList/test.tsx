import { render, screen } from 'utils/tests'
import CartList, { CartListProps } from '.'
import itemsMock from './mock'

describe('<CartList />', () => {
  const renderSut = (props?: Partial<CartListProps>) =>
    render(<CartList items={itemsMock} total="R$ 430,00" {...props} />)

  it('should render the CartList', () => {
    const { container } = renderSut()

    expect(screen.getAllByRole('heading')).toHaveLength(2)
    expect(screen.getByText('R$ 430,00')).toHaveStyle({ color: '#F231A5' })

    expect(container).toMatchSnapshot()
  })

  it('should render the button', () => {
    renderSut({ hasButton: true })

    expect(screen.getByText(/buy it now/i)).toBeInTheDocument()
  })

  it('should render empty if there are no games', () => {
    renderSut({ items: undefined })

    expect(screen.getByText(/your cart is empty/i)).toBeInTheDocument()
    expect(screen.queryByText('R$ 430,00')).not.toBeInTheDocument()
  })
})
