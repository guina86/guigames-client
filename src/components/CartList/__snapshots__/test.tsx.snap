// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> should render the heading 1`] = `
.c2 {
  padding: 1.6rem;
  border-bottom: 0.1rem solid #EAEAEA;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 1.2rem;
  width: 9.6rem;
  height: 5.6rem;
}

.c4 img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c5 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  font-size: 1.4rem;
  line-height: 1.4rem;
  font-weight: 600;
  color: #030517;
  margin-bottom: 0.8rem;
}

.c9 {
  color: #FAFAFA;
  padding: 0.2rem 1.2rem;
  background-color: #3CD3C1;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  font-weight: 600;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 100%;
}

.c0 {
  background: #FAFAFA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c1 {
  max-height: 40rem;
  overflow-y: auto;
}

.c10 {
  background: #eee;
  color: #030517;
  font-weight: 600;
  font-size: 1.4rem;
  padding: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  color: #F231A5;
}

@media (min-width:768px) {
  .c2 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    padding: 2.4rem;
  }
}

@media (min-width:768px) {
  .c4 {
    margin-right: 1.6rem;
    width: 15rem;
    height: 7rem;
  }
}

@media (min-width:768px) {
  .c7 {
    font-size: 2rem;
    line-height: 2rem;
  }
}

<div>
  <div
    class="c0"
    data-testid="cart-list"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="game-item"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <img
              alt="Red Dead Redemption 2"
              height="70"
              src="https://source.unsplash.com/user/willianjusten/151x70"
              width="150"
            />
          </div>
          <div
            class="c5"
          >
            <div
              class="c6"
            >
              <h3
                class="c7"
              >
                Red Dead Redemption 2
              </h3>
            </div>
            <div
              class="c8"
            >
              <div
                class="c9"
              >
                $215.00
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c2"
        data-testid="game-item"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <img
              alt="Borderlands 3"
              height="70"
              src="https://source.unsplash.com/user/willianjusten/151x70"
              width="150"
            />
          </div>
          <div
            class="c5"
          >
            <div
              class="c6"
            >
              <h3
                class="c7"
              >
                Borderlands 3
              </h3>
            </div>
            <div
              class="c8"
            >
              <div
                class="c9"
              >
                $215.00
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c10"
    >
      <span>
        Total
      </span>
      <span
        class="c11"
      >
        $430.00
      </span>
    </div>
  </div>
</div>
`;
